### PR TITLE
Ford: Initial support for LCA vehicles

### PR DESCRIPTION
**Pull request by incognitojam**
This PR adds support for the Focus Mk4 (EU) and Escape Mk4 using the Lane Centering function.

I am using a manual-transmission Ford Focus 2018 with stock CC and LKA ([retrofitted ACC](https://www.fordescape.org/threads/how-to-adding-adaptive-cruise-non-stop-go-and-copilot-assist.117530/) - radar seems to be factory fitted in the UK, perhaps for pre-collision assist, but ACC can be enabled using Forscan).

Thanks to Rikka, Desta, 360a and everyone else in the #ford channel for providing code, ideas and help with this. Thanks to @jyoung8607 for review and help getting this to work!

## Checklist

- [ ] added to docs/CARS.md
- [ ] test route added to [routes.py](https://github.com/commaai/openpilot/blob/master/selfdrive/car/tests/routes.py)
- [ ] route with openpilot:
- [ ] route with stock system:
- [x] car harness used (if comma doesn't sell it, put N/A): I think it is the same camera connector as the Mazda harness, but a different pinout (Molex WM10326-ND)

## Related PRs:
- panda: https://github.com/commaai/panda/pull/812
- opendbc:
  - https://github.com/commaai/opendbc/pull/476
  - https://github.com/commaai/opendbc/pull/590

### Steering commands

We can command the Power Steering Control Module (PSCM) to steer the car using the `LateralMotionControl` message, from Ford's Lane Centering Assist (LCA) / Traffic Jam Assist (TJA) feature. This gives us curvature-based control.

All cars which have LCA/TJA as an option should have a compatible PSCM fitted, and vehicles owners who didn't buy the optio from Ford should be able to enable it by changing the LCA/TJA enable bit in the PSCM As-Built Data.

From short testing so far, the `Lane_Assist_Data1` message must be sent alongside the LCA message for the PSCM to behave. This command is used by the stock system for the "Lane Keep Aid/Assist" (LKA) feature, which corrects the steering wheel when the car drifts out of the lane, but does not keep the car centered. This was the old method of commanding steering, but was subject to the 10s lockout on almost all vehicles.

Initially it was believed that `LateralMotionControl2` was used for LCA, as it has some interesting signals and checksums and counters, but then we found that we could use `LateralMotionControl`. As of yet, we don't know what `LateralMotionControl2` is used for. It is not sent on the vehicles we have tested so far. Perhaps it is used for Blue Cruise.

We can steer down to 0mph (it is not linked to the cruise control state), although the cruise control min engage speed and disengage speed varies depending on transmission and stop-and-go.

Method|Min Steer Speed|Lockout
---|---|---
LKA|35mph|Yes, 200-300ms every 10 seconds
LCA|0mph|No

Transmission|Stop and Go|Min Engage Speed|Disengage Speed
---|---|---|---
Manual|-|20mph|20mph
Automatic|No|20mph|12mph
Automatic|Yes|0mph|-

### Curvature Control

LCA is a "curvature" based control method. The Image Processing Module A (IPMA) in the vehicle writes a message describing the vehicle's current position in a lane and the lane's properties (e.g. curvature) to the CAN bus, and then the PSCM integrates this information with other data (e.g. vehicle yaw) to actuate the steering wheel to the correct position.

To achieve Lateral Control with openpilot, we can take some values from the lateral plan to estimate the lane curvature and path offset. Doing this gives adequate results, even without writing path angle and curvature rate. However, I think lane angle and curvature rate might be able to get us even better control out of the PSCM.

The signals for ramp speed and precision may be worth experimenting with when considering tuning/rate limits.

Screenshot from 2022-03-13 13-59-10

### Longitudinal Control

Although not approached in this PR, I believe we will be able to implement longitudinal control quite easily in the future. As the Cruise Control Module (C-CM, the radar) is on a private bus with the IPMA, it is the IPMA which is responsible for calculating and sending ACC commands. We can also intercept these with the same harness. We may even be able to provide ACC for vehicles without a radar, with a few coding changes.

### CODE REVIEW

**Comment by incognitojam** [comment]
I don't have stock LCA in my Ford since it's a manual-transmission and I haven't found a workaround to make it work using the As-Built Data yet.

My IPMA does output the `LateralMotionControl` messages, but does not set the mode so the PSCM does not act on it. It only sends LKA actions. I could record a route so that you could still compare this data to OP but I would be steering 🤦‍♂️.

Alternatively, another community member has an F-150 with LCA (who provided me with a route so that I could see the CAN data). Perhaps we could port the F-150 too as both a stock LCA and OP route could be provided. I think the car state implementations will be similar.

**Comment by incognitojam** [comment]
I've just rebased this branch on the latest master commit. I will continue to look at tuning when I next have free time, just it is hard to record a "drive" because the tuning is so poor at the moment I cannot let it steer for long periods, I don't know if there would be much value in sharing more drives yet. Guidances for next steps would be appreciated!

**Comment by adeebshihadeh** [comment]
> Guidances for next steps would be appreciated!

What kind of guidance are you looking for exactly? Tuning, getting this merged?

**Comment by incognitojam** [comment]
I should have been more specific :sweat_smile: I'm just racking my brains trying to figure out why this isn't steering as expected. Got some help in the #tuning channel though, and I'm going to experiment to see if the car is actually **torque** based instead of **angle**  based...

image

**Comment by incognitojam** [comment]
Screenshot from 2022-01-15 21-36-30
Currently I'm seeing the `steerRatio` increasing constantly over the length of a drive... OP steers quite strongly when only small adjustments to the steering angle are required. Will try adjusting `steerRateCost` and disabling the params learner for steer ratio.

**Comment by incognitojam** [comment]
Pinning the `steerRatio` to 1 seems to give better results with less ping pong. Current theory is that the Ford LCA command accepts the _wheels_ steering angle rather than the steering wheel angle.

**Comment by jyoung8607** [comment]
> Pinning the `steerRatio` to 1 seems to give better results with less ping pong. Current theory is that the Ford LCA command accepts the _wheels_ steering angle rather than the steering wheel angle.

Looking at your DBC, you were right in your opening comment that it's literally a path angle (plus a curvature, and some other things). That ends up being important.

https://github.com/incognitojam/opendbc/blob/192a11001cfa3047ebc19b42519c65f70c08ec5e/ford_lincoln_base_pt.dbc#L3365-L3374

Normally, openpilot figures out the curvature and curvature rate of the path it wants to follow, and then computes a desired _steering_ angle to follow that path based on the vehicle model (wheelbase, steering ratio), and then computes actuators to make that happen, either by passing along that angle to EPS which consume it directly, or computing a torque feedback loop to indirectly effect that same steering angle.

We can infer from your DBC that your IPMA-to-PSCM abstraction is different. The LateralMotionControl contains a _path_ angle (as opposed to steering angle), a curvature, a curvature rate, an offset from the center of that path, and how aggressive it should be about following it. The implication is that your IPMA sees a road it wants to follow, describes that road to the the PSCM, and otherwise says "pls just handle it kthxbai".

Bad news, openpilot isn't really set up to do that. The car port abstractions are designed to receive pre-computed actuators. Good news, making openpilot do that shouldn't be rocket surgery. Teach controlsd about a new `curvature` entry for [`SteerControlType`](https://github.com/commaai/cereal/blob/d61741abed022e6f5f5ce7614d0d3b0d552258ef/car.capnp#L544-L547) and  [`LateralControlState`](https://github.com/commaai/cereal/blob/d61741abed022e6f5f5ce7614d0d3b0d552258ef/log.capnp#L574-L580), add a new `LateralCurvatureState`, figure out a clean way to pass the existing [desired curvature, curvature rate](https://github.com/commaai/openpilot/blob/59894a66669f72dd65acf8f7556c43007d1af68e/selfdrive/controls/controlsd.py#L499-L506), and maybe [path deviation](https://github.com/commaai/openpilot/blob/59894a66669f72dd65acf8f7556c43007d1af68e/selfdrive/controls/controlsd.py#L534-L539) down to the car port from the upper layer of controlsd. Saturation and limit checks somewhere, probably have to tell the params learner ssh bby is okay when in curvature mode, and of course Panda safety code.

Tagging @HaraldSchafer and @pd0wm for visibility and comment. If supporting Ford LCA requires tinkering with some very old and very well understood abstractions in controlsd, they should be consulted to make sure the solution is upstreamable.

**Comment by jyoung8607** [comment]
Since your vehicle doesn't have LCA, do you have anyone that can share an uploaded route from a vehicle that does? 

I've been trying to figure out what path angle means in this context.  I think `LatCtlPath_An_Actl` might be complementary to `LatCtlPathOffst_L_Actl`, to describe how far off the vehicle is from the IPMA planned path, in yaw angle and lateral distance respectively.

Would be interesting to have a driver engage LCA on a highway with a long curve, take the wheel, and hold the vehicle a little to one side of the lane. If I'm right, you would show a little bit of `LatCtlPathOffst_L_Actl` lateral offset, but since you're still paralleling the requested path, I suspect `LatCtrlPath_An_Actl` would stay fairly close to zero.

**Comment by incognitojam** [comment]
Thanks for the analysis @jyoung8607, it's great that you're piecing the puzzle together!

It's late here so I'll have to give this another read in the morning, but I thought I could respond about the lane centering data/recorded routes.

The IPMA always outputs the lane centering messages, even when it isn't engaged and even though it is not configured to work in my car! The only signal that isn't set is `LatCtl_D_Rq` from what I can tell.

However, 360a#3201 has been sharing routes with me to show how stock LCA works as well as testing this branch out. I'm sure he could share something with you!

**Comment by jyoung8607** [comment]
Talked to Team Ford on Discord tonight. They're going to get some highway routes in dashcam/stock LCA mode tomorrow, and we'll see what they look like in PlotJuggler. With stock LCA engaged and conforming well (stable in center of lane) along a curve, significant activity (or lack thereof) in `LatCtlPath_An_Actl` should tell us if it's really wants to drive the curvature instead.

**Comment by jyoung8607** [comment]
image

Taken from a middle chunk of `85ad625d9a86b88d|2022-01-18--07-01-11`. From top to bottom:

* Lateral path offset (meters) from IPMA
* Lateral path angle (radians) from IPMA
* openpilot CarState steering angle (degrees), overlaid with stock LCA/IPMA engaged state
* Lateral curvature (1/meters) from IPMA

Stock LCA clearly isn't using `LatCtlPath_An_Actl` as a direct way to drive the steering angle. The units are different, obviously, but the car's steering angle is following the general waveform of the IPMA curvature, and not really tracking the IPMA path angle. I think it's operating as I described above, the difference in yaw needed to align to the desired path. 

Where you do see path angle reactions, it includes a path offset distance as well. See around 280 +/- for what looks like a reaction to a disturbance/bump, and around 365 for a lane change. Particularly for the lane change, note how the path offset snaps over to the new lane, and the path angle points right toward it, and then reduces to near zero, but exhibits none of the "ringing" waveform/S-curve you see in the vehicle's steering angle as the vehicle straightens up in its new lane.

**TL;DR**: this thing wants a curvature and curvature rate.

As a quick and dirty test, I'd lock path angle and path offset to zero (effectively telling it "you're exactly on the desired path") and then feed it `desired_curvature` and `desired_curvature_rate` and see if it goes down the road without catching on fire. I _think_ this will work? openpilot doesn't do lane-changes and the like by saying "the path is now over -->there", it plots a path from where you're at to where it wants to be.

I am swimming well beyond my usual depth in controlsd, so feedback from comma or other controls-smart people would be highly welcome.

**Comment by incognitojam** [comment]
I've written some hacky code re-using `actuators.steer` to store the curvature rate, and set path angle and path offset to zero. I can't test it right now but will be able to tomorrow.

**Comment by incognitojam** [comment]
> The car port interfaces are designed to use pre-computed actuators. Good news, making openpilot do that shouldn't be rocket surgery. Teach controlsd about a new curvature entry for SteerControlType and LateralControlState, add a new LateralCurvatureState, figure out a clean way to pass the existing desired curvature, curvature rate, and maybe path deviation down to the car port from the upper layer of controlsd.

I've pushed a very messy first attempt at this, and after doing it I think it might be better to create a new struct for curvature controls instead of shoving everything into `Actuators`, and pass it inside `CarControl`.

**Comment by incognitojam** [comment]
Just trying to figure out why there is a merge conflict :thinking:

**Comment by incognitojam** [comment]
Some really good results earlier tonight!

I'm using the `curvature` and `curvatureRate` values from `lateralPlan` instead of the "desired curvature", as the Ford PSCM wants to know the curvature of the _road_, not of the current maneouvre OP is doing, and we pass additional information about the path offset/angle anyway.

After reviewing a drive using the `desired_curvature` in PlotJuggler, I could see the ping pong effect being introduced as the PSCM thought the road was curving when actually OP was sending the curvature of the corrective action it was taking in the planner. If instead we send a curvature further down the path, where the path closely matches the road curvature, it seems to work well. I chose `lat_plan.curvatures[8]` and `lat_plan.curvatureRates[15]` after some experimentation.

We are effectively sending the future curvature but it still works OK. I've found it to be really good on laneless.

It's kind of hacky the way I've implemented it, but someone with more knowledge about controlsd and the lateralplanner might be able to suggest a better method!

**Comment by incognitojam** [comment]
Current issues:
- Params learner gets confused about the steering angle offset and throws a vehicle parameters alert. I think this is because OP interally knows what angle we should steer to in order to follow its lateral plan, but we aren't sending that to the car, we're sending the curvature/path deviation etc. and allowing the PSCM to steer to follow the path. So when the PSCM does something differently to what OP expects there is a large offset between the intended steering angle and the actual angle.

- 360a reported that cruise is cancelled below 15mph, unlike with stock. I'm not sure why this happens yet. I can't test this on my car as I don't have automatic transmission/stop and go.

- OP is changing lane with the turn signal on without a nudge. I think this will be an issue with steering torque values recorded in car state.

**Comment by incognitojam** [comment]
Just rebased on top of master after the latcontrol PR was merged

**Comment by incognitojam** [comment]
Curvature control actuators, quick run-down for where we are right now: ([link to curvature control](https://github.com/commaai/openpilot/pull/23331/files#diff-4beea7cc7c1fb318eab2a0251db8a8c276ab77f1afca876004e03531a9398357R7))

I am using `lateralPlan.dPathPoints` for `actuators.pathDeviation`. I verified this closely matches the shape of the signal from the stock system (TODO: screenshot from plotjuggler).

I was originally using `desired_curvature` as `actuators.curvature`, but this produced oscillations/a ping pong effect. I think this was caused by the `desired_curvature` being calculated from OP's planned path, which at the beginning includes any manoeuvure to move to the centre of the lane (or to the desired position in the lane). However, since we send the "path deviation" to the PSCM too, the PSCM does additional calculations with this and in effect applies the manoeuvre twice, once from the manipulated curvature (by OP) and second by the path deviation signal.

To fix this I just chose a "later" curvature value from `lateralPlan.curvatures`, since I thought this may be closer to the "road curvature" where we are centred in the lane. This is why I chose `lateralPlan.curvatures[8]`, and from simple analysis in PlotJuggler to verify it was similar to the stock values (TODO: screenshot from plotjuggler) and manual testing it seems to work well on straight roads. However, on curves, I think the car is responding too soon. This is where I think I'll need help finding a better solution :)

I haven't spent the time to investigate `actuators.curvatureRate` yet. `lateralPlan.curvatureRates[15]` looked similar in magnitude to the stock LCA signal so I chose that. This might be a dumb choice :sweat_smile: 

For `actuators.pathAngle`, I'm not sure where to begin to find this in the lateral plan. I believe this is the angle between the car's heading and the desired path. Will this have to be created new? I'm using `0` for now and it still provides satisfactory results.

**Comment by incognitojam** [comment]
Just created a separate PR for the "try Panda bus for FPv2 queries" stuff ^^

**Comment by incognitojam** [comment]
Added some signals for `yawRate`, `brake` (although it's brake torque? not pedal pct) and I think I've found a better signal for `steeringPressed`, too

**Comment by EasternPA** [comment]
Following with great excitement. Any reason why the suggested documentation update in the first post indicates that Automatics with Stop and Go will disengage at 10 MPH? What am I missing?

**Comment by incognitojam** [comment]
> Following with great excitement. Any reason why the suggested documentation update in the first post indicates that Automatics with Stop and Go will disengage at 10 MPH? What am I missing?

I think I might've made a mistake with that actually, was just from memory, it doesn't make sense actually!

**Comment by scripty29** [comment]
Hey @incognitojam i have been following you for a little while. I have a 2019 F150 with ACC stop and go and LKS. Any chance your code will work with my truck and or do you need additional testers? Also willing to learn how to tune etc.. but I don’t have a lot of coding experience. Cheers on the work you have done so far!

**Comment by incognitojam** [comment]
> I have a 2019 F150 with ACC stop and go and LKA. Any chance your code will work with my truck

Hi @scripty29 , I'm not sure it will work on your F150... Do you know if Ford offered "traffic jam assist" or lane centering option from the factory? You might not have a compatible power steering module

You'd be able to check if you get an OBD adapter and the Forscan software

**Comment by scripty29** [comment]
I have Forscan and can check this afternoon! I will report back with my findings 🤞

**Comment by incognitojam** [comment]
Just rebased this onto the latest master, using the new FW query stuff so that there isn't a hack for choosing the panda bus anymore

File: selfdrive/car/ford/values.py [hunk 1]
```diff
@@ -1,21 +1,98 @@
...
+  camera = 2
 
 
 class CAR:
-  FUSION = "FORD FUSION 2018"
+  ESCAPE_MK4 = "FORD ESCAPE 4TH GEN"
+  FOCUS_MK4 = "FORD FOCUS 4TH GEN"
+
+
+class Footnote(Enum):
```
Comment (adeebshihadeh) [inline selfdrive/car/ford/values.py:36]: no need for any of this docs stuff yet. we'll add once they're out of dashcam only

**Review by adeebshihadeh — COMMENTED** [review]
I'd like to get this merged as it seems very similar to the current F150 platform. Here's some quick comments after going through the PR - I'll do a thorough review sometime this week. 

Let's start by whittling this down to a subset of this PR that can be merged right away:
* fordcan changes to build the new messages
* carState
* carController
* new car models + fusion removal
* possibly the FW query
* re-enable dashcamOnly

Feel free to do that in this PR or break out into a new one.

File: selfdrive/car/ford/radar_interface.py [hunk 2]
```diff
@@ -1,72 +1,5 @@
 #!/usr/bin/env python3
-from cereal import car
-from common.conversions import Conversions as CV
-from opendbc.can.parser import CANParser
-from selfdrive.car.ford.values import DBC
 from selfdrive.car.interfaces import RadarInterfaceBase
 
-RADAR_MSGS = list(range(0x500, 0x540))
```
[the remark this answers is not in the document]
Comment (incognitojam) [reply selfdrive/car/ford/radar_interface.py:8]: I'll have to test this again... I can't remember what the problem was

File: selfdrive/car/ford/interface.py [hunk 3]
```diff
@@ -1,61 +1,74 @@
...
-    ret.mass = 3045. * CV.LB_TO_KG + STD_CARGO_KG
-    ret.lateralTuning.pid.kiBP, ret.lateralTuning.pid.kpBP = [[0.], [0.]]
-    ret.lateralTuning.pid.kpV, ret.lateralTuning.pid.kiV = [[0.01], [0.005]]     # TODO: tune this
-    ret.lateralTuning.pid.kf = 1. / MAX_ANGLE   # MAX Steer angle to normalize FF
-    ret.steerActuatorDelay = 0.1  # Default delay, not measured yet
-    ret.steerLimitTimer = 1.0
+    ret.radarOffCan = True
+
+    # Curvature-based steering
+    ret.steerControlType = car.CarParams.SteerControlType.curvature
```
[the remark this answers is not in the document]
Comment (incognitojam) [reply selfdrive/car/ford/interface.py:22]: opened https://github.com/commaai/openpilot/pull/24181

File: selfdrive/car/ford/interface.py [hunk 4]
```diff
@@ -1,61 +1,74 @@
...
-    # TODO: add minSteerSpeed
+
+    ret.minSteerSpeed = 0.  # LCA works down to zero
     ret.minEnableSpeed = 12. * CV.MPH_TO_MS
 
-    # TODO: get actual value, for now starting with reasonable value for
-    # civic and scaling by mass and wheelbase
-    ret.rotationalInertia = scale_rot_inertia(ret.mass, ret.wheelbase)
+    # Auto Transmission: Gear_Shift_by_Wire_FD1
+    if 0x5A in fingerprint[0]:
```
[the remark this answers is not in the document]
Comment (incognitojam) [reply selfdrive/car/ford/interface.py:34]: Is there an example of how to do this somewhere else?

[the remark this answers is not in the document]
Comment (incognitojam) [reply selfdrive/car/ford/interface.py:34]: actually, I found how to do it by reading `car_fw`, but I guess you have to do this after the FPv2 stuff is added

File: selfdrive/car/ford/carstate.py [hunk 5]
```diff
@@ -1,58 +1,222 @@
...
+    elif self.CP.transmissionType == TransmissionType.manual:
+      gear = cp.vl["PowertrainData_10"]["GearPos_D_Actl"]
+      if gear == 0:
+        ret.gearShifter = GearShifter.neutral
+      elif gear in (1, 2, 3, 4, 5, 6, 7, 8, 9, 10):
+        ret.gearShifter = GearShifter.drive
+      elif gear == 14:
+        ret.gearShifter = GearShifter.reverse
+      else:
+        ret.gearShifter = GearShifter.unknown
```
[the remark this answers is not in the document]
Comment (incognitojam) [reply selfdrive/car/ford/carstate.py]: I've attempted this and opened a PR to opendbc, but I think I need to delete info from the existing DBC and replace with the new simplified table to make this work.. Is there a way to preserve both? https://github.com/commaai/opendbc/pull/590

File: selfdrive/car/ford/interface.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (incognitojam) [reply selfdrive/car/ford/interface.py:47]: From memory, the transmission ECU never appeared in the car_fw... Only saw the eps, esp, radar, camera and engine responses.

File: selfdrive/test/process_replay/test_processes.py [hunk 6]
```diff
@@ -47,7 +47,7 @@
 ]
 
 # dashcamOnly makes don't need to be tested until a full port is done
-excluded_interfaces = ["mock", "ford", "mazda", "tesla"]
+excluded_interfaces = ["mock", "mazda", "tesla"]
```
Comment (adeebshihadeh) [inline selfdrive/test/process_replay/test_processes.py:50]: add this back

**Comment by adeebshihadeh** [comment]
Nice job @incognitojam! Seems we'll be able to support quite a few existing Fords as well as the new Blue Cruise platform with this port. `LateralMotionControl2` is present on the F-150 with Blue Cruise, but we never saw it steer due to the faults.

Here's the next steps:
* break out FW query into a new PR
* finish & merge the matching panda safety
* good quality steering with the curvature message

We can also review and merge the matching panda safety #812 once it's ready, though it'll stay behind the debug panda flag until we get a Ford again and verify it. We'll be ready to invest more time into this once the whole port is working well on a car from the factory.
